### PR TITLE
New simply script path

### PR DESCRIPTION
**Pull request by lxxxvi**
@nerdinand 

The `simpli` gem should now be able to call the simply (this one) script. 

2 changes are required:
* The `simply` script should be available at `./bin/simply.rb`
* changes in this PR

Just a proposal

### CODE REVIEW

**Review by nerdinand — COMMENTED** [review]
I'm not sure about this... In the worst case, the user has to type this. `./simply foo` is much shorter than `./bin/simply.rb foo`. What do you think, @phylor @pic ?

**Comment by phylor** [comment]
@nerdinand @lxxxvi Is `.rb` required? `bin/simply` would match with all the other scripts in `bin`.

I like the gem idea. Although I'm really confused now. Why do we need a script if we already have a gem? We could include the `simply` script in the gem and just add it to our `Gemfile`'s `:development`, right? It would then be available immediately after checking out a project and bundling it.

And if you want to do anything before bundling, you could still `gem install simpli`. Rails has the same problem: if you want to create a new project, you cannot bundle, because, oh well, you don't have a project. So you need to install Rails globally.

**Comment by phylor** [comment]
@nerdinand @lxxxvi @pic Sorry for being annoying, but are there any updates/comments on this? I'd like to use it, but wonder how (btw, we should update the `README` with installation instructions).

Do we like the gem idea? Should we move the script into the gem?

**Comment by nerdinand** [comment]
@phylor It's not annoying, no worries.

I can't sleep so I'm writing this answer in the middle of the night... :-P

Here's my take on the gem idea: The `simpli` gem, as implemented by @lxxxvi only does the following: It allows running `simply` in any directory and will automatically find the correct script to run (i.e. when you're in a subdirectory within your project it will traverse the directory structure upwards until it finds `simply`).
The way I usually work, I have a terminal open at the root of the project anyway, so for me this gem is a bit pointless. It doesn't do any harm having it though, so if you like it, you might as well use it imho. ;-)

Furthermore, in my opinion there are some reasons that speak against having a gem implementation of the `simply` script:

- We have projects that are not Rails (or even Ruby) projects and thus have no `Gemfile`. To add a `Gemfile` to these projects just to have `simply` is overkill in my opinion.
- You might want to do things to set up a project before actually bundling, e.g. installing dependencies for gems (take blue-space as an example, where I install `qt`: https://github.com/simplificator/blue-space/blob/master/bin/simply/setup). For that to work, the `simply` script should be absolutely self-contained.

In summary, I quite like the self-contained nature of the `simply` script. That was a requirement and design goal from the beginning. And I like the idea of only having to run `./simply setup` to be up and running in a new project with just a few things you need to have installed beforehand:
- `ruby` (comes with macOS anyway)
- `rbenv` (if the project is a Ruby project)
- `brew` (who doesn't want that installed anyway?)

@phylor I'll update the `README` tomorrow!

**Comment by lxxxvi** [comment]
@phylor 
@nerdinand is right. The `simpli` gem only does the upsearch to the "actual" simply script. It's optional and auxiliary. Functionalities are not supposed to go into the gem for the reasons he pointed out.

However, my personal preference is, and many of you may disagree, to have 

a) file extensions for every non-directory files (except executables in $PATH)
b) files in a proper places

That's why I proposed to move the `./simply` script to `./bin/simply.rb`(rule a) and b)) and have the related scripts in the directory `./bin/simply/` (rule b), already compliant).

Don't worry too much about the `simpli` gem. 

> The way I usually work, I have a terminal open at the root of the project anyway, so for me this gem is a bit pointless.

Yes, it is pointless, as long as the `./simply` script is in the root directory (aka the wrong directory, see rule b)) ;-)

Btw, you could also create an `alias simply=./bin/simply.rb` to have the convenience in the root directory.

**Comment by nerdinand** [comment]
I just think in order to make adoption as smooth as possible, we should minimise the amount of work for the developer and maximise convenience. In my opinion my solution provides that. All you have to do as a dev is to `git clone` the repository and run `./simply setup`. It really does not get any simpler as that.

**Comment by lxxxvi** [comment]
@nerdinand 
I understand. What if a project (needs to have|has) a folder named `./simply/` in the root directory... ;-)

Just kidding. It's just a proposal, so I suggest you close the PR ;-)

**Comment by phylor** [comment]
@lxxxvi I'd still appreciate partial changes:

- I don't like the project folders to be convoluted and I agree that executables should be in `bin` (as that's the convention). I think it's a good idea to move the script to `bin` and run it using `bin/simply`. As we already do with `bin/rubocop`, `bin/rails`, etc. But I would not add the file extension, because we have a hashbang in the script. If you were to call it using `ruby simply.rb`, add the extension, but `bin/simply` should be used without IMO.
- I fear that we have to force people to upgrade the script whenever it changes. With a gem that would be easier. I also don't like to include deployment scripts/helpers in the code repository as that violates the 12 factor app. But I can very well live with that (and I don't have a good solution alternative or compelling arguments).

@lxxxvi @nerdinand Would that be a good compromise? To move the script to `bin/simply`?

**Comment by lxxxvi** [comment]
@phylor 

OK, now you're thinking ahead. I can't handle this. ;-)

>  I think it's a good idea to move the script to bin and run it using bin/simply. As we already do with bin/rubocop, bin/rails, etc.

Now you have the problem, that in `./bin/` you have a folder named `simply` for the actual scripts. That means you cannot have a `./bin/simply` script there. 

> I fear that we have to force people to upgrade the script whenever it changes. 

This is something that, IMO, the gem could do. The gem could provide "simply init" to curl the scripts and put them into the right places. But that is thinking ahead. :-)

**Comment by phylor** [comment]
Silly me, let's have `bin/simply` and `bin/simplyrc/` (or `bin/simply-scripts/` or `bin/simply-bin/`). What do you think? :)

Good idea about the gem! `simply init` is already taken by the script though. What about `simply install` and `simply upgrade`? Could be confused with the script though, so maybe we need a `simply script-install`.

**Comment by nerdinand** [comment]
I like the compromise of putting it into `bin/simply`. I'm also down with using the gem to upgrade the script. I've made this project public now on github to facilitate installation so upgrading it is as simple as running `curl https://raw.githubusercontent.com/simplificator/simply/master/simply > simply`.

File: simpy.rb [hunk 1]
```diff
@@ -167,6 +171,7 @@ def usage
   end
 
   def perform
+    puts "PERFORM TEST #{executable_path}"
```
Comment (nerdinand) [inline simpy.rb]: BTW, what's the purpose of this line?

**Comment by nerdinand** [comment]
@phylor @lxxxvi I made some changes, can we agree to this?

**Comment by nerdinand** [comment]
Another "problem" just occurred to me: We're always thinking about Rails projects here. What if the project you add `simply` to doesn't have a `bin` directory or that directory serves another purpose? E.g. compiled binaries?

**Comment by lxxxvi** [comment]
@phylor @nerdinand 

🇨🇭 like compromises... 👍

File: simply [hunk 2]
```diff
@@ -167,6 +171,7 @@ class TestCommand < Command
   end
 
   def perform
+    puts "PERFORM TEST #{executable_path}"
```
Comment (phylor) [inline simply]: Could we remove this line?

**Review by phylor — COMMENTED** [review]
Looks perfect! I'd remove the `puts` though.

File: simply (hunk 2 above)
Comment (nerdinand) [reply simply]: Yes, I agree. What say you, @lxxxvi?

**Comment by phylor** [comment]
@nerdinand This works:

```
cd /tmp
mkdir testproj
cd testproj
curl https://raw.githubusercontent.com/simplificator/simply/new-simply-script-path/simply > simply
./simply init

ls -lR
total 16
-rwxr-xr-x  1 serge  wheel  4815 Mar  1 11:59 simply
drwxr-xr-x  7 serge  wheel   238 Mar  1 11:59 simply-executables

./simply-executables:
total 40
-rwxr-xr-x  1 serge  wheel  241 Mar  1 11:59 build
-rwxr-xr-x  1 serge  wheel  245 Mar  1 11:59 deploy
-rwxr-xr-x  1 serge  wheel  233 Mar  1 11:59 run
-rwxr-xr-x  1 serge  wheel  241 Mar  1 11:59 setup
-rwxr-xr-x  1 serge  wheel  237 Mar  1 11:59 test
```

We could check in the gem if there is a bin directory and if there is not, just install the script into the root of the project. Or ask the user.

**Comment by nerdinand** [comment]
@phylor Honestly, I'd just leave it as is... I think the user should read the README and the rest is not our problem... ;-)

File: simply (hunk 2 above)
Comment (lxxxvi) [reply simply]: fine for me
